File: tours_and_travels/tours_and_travels/report/guide_wise_tour_registration_analytics/guide_wise_tour_registration_analytics.js
// Copyright (c) 2022, Nihantra C. Patel and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Guide-wise Tour Registration Analytics"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_start_date"),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.defaults.get_user_default("year_end_date"),
            "reqd": 1
        },
        {
            "fieldname": "name",
            "label": __("Tour Registration"),
            "fieldtype": "MultiSelectList",
            "options": "Tour Registration",
            get_data: function(txt) {
                return frappe.db.get_link_options('Tour Registration', txt);
            }
        },
        {
            "fieldname": "tour_package",
            "label": __("Tour Package"),
            "fieldtype": "MultiSelectList",
            "options": "Tour Package",
            get_data: function(txt) {
                return frappe.db.get_link_options('Tour Package', txt);
            }
        },
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "MultiSelectList",
            "options": "Customer",
            get_data: function(txt) {
                return frappe.db.get_link_options('Customer', txt);
            }
        },
        {
            "fieldname": "package_category",
            "label": __("Package Category"),
            "fieldtype": "MultiSelectList",
            "options": "Package Category",
            get_data: function(txt) {
                return frappe.db.get_link_options('Package Category', txt);
            }
        },
        {
            "fieldname": "package_type",
            "label": __("Package Type"),
            "fieldtype": "Select",
            "options": ["", "Domestic", "International"]
        },
        {
            "fieldname": "state",
            "label": __("State"),
            "fieldtype": "MultiSelectList",
            "options": "State",
            get_data: function(txt) {
                return frappe.db.get_link_options('State', txt);
            }
        },
        {
            "fieldname": "country",
            "label": __("Country"),
            "fieldtype": "MultiSelectList",
            "options": "Country",
            get_data: function(txt) {
                return frappe.db.get_link_options('Country', txt);
            }
        },
        {
            "fieldname": "show_package",
            "label": __("Show Attraction Package"),
            "fieldtype": "Check"
        },
        {
            "fieldname": "guide",
            "label": __("Guide"),
            "fieldtype": "MultiSelectList",
            "options": "Guide",
            "depends_on": "show_package",
            get_data: function(txt) {
                return frappe.db.get_link_options('Guide', txt);
            }
        },
        {
            "fieldname": "guide_service",
            "label": __("Guide Service"),
            "fieldtype": "MultiSelectList",
            "options": "Guide Service",
            "depends_on": "show_package",
            get_data: function(txt) {
                return frappe.db.get_link_options('Guide Service', txt);
            }
        },
        {
            "fieldname": "city",
            "label": __("City"),
            "fieldtype": "MultiSelectList",
            "options": "City",
            "depends_on": "show_package",
            get_data: function(txt) {
                return frappe.db.get_link_options('City', txt);
            }
        }
    ],

    onload: function(report) {
        report.page.set_inner_btn_group_as_primary(__('Tour Registration Analytics'));

        report.page.add_inner_button(("Hotel-wise Tour Registration Analytics"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Hotel-wise Tour Registration Analytics', {
                company: filters.company,
                from_date: filters.from_date,
                to_date: filters.to_date,
                name: filters.name,
                tour_package: filters.tour_package,
                package_category: filters.package_category,
                package_type: filters.package_type,
                city: filters.city,
                state: filters.state
            });
        }, __('Tour Registration Analytics'));

        report.page.add_inner_button(("Restaurant-wise Tour Registration Analytics"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Restaurant-wise Tour Registration Analytics', {
                company: filters.company,
                from_date: filters.from_date,
                to_date: filters.to_date,
                name: filters.name,
                tour_package: filters.tour_package,
                package_category: filters.package_category,
                package_type: filters.package_type,
                city: filters.city,
                state: filters.state
            });
        }, __('Tour Registration Analytics'));

        report.page.add_inner_button(("Transportation-wise Tour Registration Analytics"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Transportation-wise Tour Registration Analytics', {
                company: filters.company,
                from_date: filters.from_date,
                to_date: filters.to_date,
                name: filters.name,
                tour_package: filters.tour_package,
                package_category: filters.package_category,
                package_type: filters.package_type,
                city: filters.city,
                state: filters.state
            });
        }, __('Tour Registration Analytics'));

        report.page.add_inner_button(("Attraction-wise Tour Registration Analytics"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Attraction-wise Tour Registration Analytics', {
                company: filters.company,
                from_date: filters.from_date,
                to_date: filters.to_date,
                name: filters.name,
                tour_package: filters.tour_package,
                package_category: filters.package_category,
                package_type: filters.package_type,
                city: filters.city,
                state: filters.state
            });
        }, __('Tour Registration Analytics'));

        report.page.add_inner_button(("Tour Registration Analytics"), function() {
            var filters = report.get_values();
            frappe.set_route('query-report', 'Tour Registration Analytics', {
                company: filters.company,
                from_date: filters.from_date,
                to_date: filters.to_date,
                name: filters.name,
                tour_package: filters.tour_package,
                package_category: filters.package_category,
                package_type: filters.package_type,
                city: filters.city,
                state: filters.state
            });
        }, __('Tour Registration Analytics'));
    }
};

File: tours_and_travels/tours_and_travels/report/guide_wise_tour_registration_analytics/guide_wise_tour_registration_analytics.py
# Copyright (c) 2022, Nihantra C. Patel and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	columns, data = get_columns(filters), get_data(filters)
	return columns, data

def get_columns(filters):
	columns = [
			{
				"label": _("Date"),
				"fieldname": "date",
				"fieldtype": "Date",
				"width": 100
			},
			{
				"label": _("Tour Regi. ID"),
				"fieldname": "name",
				"fieldtype": "Link",
				"options": "Tour Registration",
				"width": 110
			},
			{
				"label": _("Tour Package"),
				"fieldname": "tour_package",
				"fieldtype": "Link",
				"options": "Tour Package",
				"width": 110
			},
			{
				"label": _("Customer"),
				"fieldname": "customer",
				"fieldtype": "Link",
				"options": "Customer",
				"width": 120
			},
		]

	if filters.get("show_package") != 1:
		columns.append(
			{
				"label": _("Guide Total Amount"),
				"fieldname": "guide_total_amount",
				"fieldtype": "Currency",
				"width": 140
			})

	if filters.get("show_package") == 1:
		columns.append(
			{
				"label": _("Date"),
				"fieldname": "date",
				"fieldtype": "Date",
				"width": 100
			})
		columns.append(
			{
				"label": _("City"),
				"fieldname": "city",
				"fieldtype": "Link",
				"options": "City",
				"width": 100
			})
		columns.append(
			{
				"label": _("Guide"),
				"fieldname": "guide",
				"fieldtype": "Link",
				"options": "Guide",
				"width": 200
			})
		columns.append(
			{
				"label": _("Guide Service"),
				"fieldname": "guide_service",
				"fieldtype": "Link",
				"options": "Guide Service",
				"width": 120
			})
		columns.append(
			{
				"label": _("Contact"),
				"fieldname": "contact",
				"fieldtype": "Data",
				"width": 120
			})
		columns.append(
			{
				"label": _("Price"),
				"fieldname": "price",
				"fieldtype": "Currency",
				"width": 100
			})

	columns.extend(
		[
			{
				"label": _("Package Name"),
				"fieldname": "package_name",
				"fieldtype": "Data",
				"width": 140
			},
			{
				"label": _("Package Type"),
				"fieldname": "package_type",
				"fieldtype": "Data",
				"width": 115
			},
			{
				"label": _("Package Category"),
				"fieldname": "package_category",
				"fieldtype": "Link",
				"options": "Package Category",
				"width": 150
			},
			{
				"label": _("Arrival Date"),
				"fieldname": "arrival_date",
				"fieldtype": "Date",
				"width": 100
			},
			{
				"label": _("Departure Date"),
				"fieldname": "departure_date",
				"fieldtype": "Date",
				"width": 100
			},
			{
				"label": _("State"),
				"fieldname": "state",
				"fieldtype": "Link",
				"options": "State",
				"width": 100
			},
			{
				"label": _("Country"),
				"fieldname": "country",
				"fieldtype": "Link",
				"options": "Country",
				"width": 100
			}
		]
		)
	return columns

def get_data(filters):
	if filters.get("show_package") == 1:
		return frappe.db.sql(
				"""
				SELECT
					tr.date, tr.name, tr.tour_package, tr.customer, tr.guide_total_amount,
					gp.date, gp.city, gp.guide, gp.guide_service, gp.contact, gp.price, tr.package_name,
					tr.package_type, tr.package_category, tr.arrival_date, tr.departure_date, tr.state, tr.country

				FROM
					`tabTour Registration` as tr, `tabGuide Package List` as gp

				WHERE
					gp.parent = tr.name
					AND tr.docstatus < 2
					AND tr.date BETWEEN %(from_date)s AND %(to_date)s

					{conditions}

				ORDER BY
					tr.date DESC """.format(
					conditions=get_conditions(filters)
				),
				filters,
				as_dict=1,
			)
	else:
		return frappe.db.sql(
				"""
				SELECT
					tr.date, tr.name, tr.tour_package, tr.customer, tr.guide_total_amount, tr.package_name,
					tr.package_type, tr.package_category, tr.arrival_date, tr.departure_date, tr.state, tr.country

				FROM
					`tabTour Registration` as tr

				WHERE
					tr.docstatus < 2
					AND tr.date BETWEEN %(from_date)s AND %(to_date)s

					{conditions} """.format(
					conditions=get_conditions(filters)
				),
				filters,
				as_dict=1,
			)		

def get_conditions(filters):
	conditions = []

	if filters.get("company"):
		conditions.append(" AND tr.company = %(company)s")

	if filters.get("name"):
		conditions.append(" AND tr.name in %(name)s")

	if filters.get("tour_package"):
		conditions.append(" AND tr.tour_package in %(tour_package)s")

	if filters.get("customer"):
		conditions.append(" AND tr.customer in %(customer)s")

	if filters.get("package_category"):
		conditions.append(" AND tr.package_category in %(package_category)s")

	if filters.get("package_type"):
		conditions.append(" AND tr.package_type = %(package_type)s")

	if filters.get("city"):
		conditions.append(" AND gp.city in %(city)s")

	if filters.get("state"):
		conditions.append(" AND tr.state in %(state)s")

	if filters.get("country"):
		conditions.append(" AND tr.country in %(country)s")

	if filters.get("guide"):
		conditions.append(" AND gp.guide in %(guide)s")

	if filters.get("guide_service"):
		conditions.append(" AND gp.guide_service in %(guide_service)s")

	return " ".join(conditions) if conditions else ""